test/jdk/tools/launcher/modules/patch/automatic/PatchTest.java

```diff
@@ -0,0 +1,173 @@
+
+public class PatchTest {
+
+    private static final String APP_NAME = "myapp";
+
+    private static final String MODULE_NAME = "somelib";
+
+    private static final String PATCH1_NAME = "patch1";
+    private static final String PATCH2_NAME = "patch2";
+
+    private static final String APP_MAIN = "myapp.Main";
+    private static final String PATCH1_MAIN = "somelib.test.TestMain";
+    private static final String PATCH2_MAIN = "somelib.Dummy";
+
+    private static final String TEST_SRC = System.getProperty("test.src");
+
+    private static final Path APP_SRC = Path.of(TEST_SRC, APP_NAME);
+    private static final Path APP_CLASSES = Path.of("classes", APP_NAME);
+    private static final Path SOMELIB_SRC = Path.of(TEST_SRC, MODULE_NAME);
+    private static final Path SOMELIB_PATCH1_SRC = Path.of(TEST_SRC, PATCH1_NAME);
+    private static final Path SOMELIB_PATCH2_SRC = Path.of(TEST_SRC, PATCH2_NAME);
+    private static final Path SOMELIB_CLASSES = Path.of("classes", MODULE_NAME);
+    private static final Path SOMELIB_PATCH1_CLASSES = Path.of("classes", PATCH1_NAME);
+    private static final Path SOMELIB_PATCH2_CLASSES = Path.of("classes", PATCH2_NAME);
+    private static final Path SOMELIB_JAR = Path.of("mods", MODULE_NAME + "-0.19.jar");
+
+    private static final String MODULE_PATH = String.join(File.pathSeparator, SOMELIB_JAR.toString(), APP_CLASSES.toString());
+
+    /**
+     * The test consists of 2 modules:
+     *
+     * somelib - dummy automatic module.
+     * myapp - explicit module, uses somelib
+     *
+     * And two patches:
+     *
+     * patch1 - adds an additional package.
+     * patch2 - only replaces existing classes.
+     *
+     */
+    @BeforeClass
+    public void compile() throws Exception {
+        boolean compiled;
+
+        // create mods/somelib-0.19.jar
+
+        compiled = CompilerUtils.compile(SOMELIB_SRC, SOMELIB_CLASSES);
+        assertTrue(compiled);
+
+        JarUtils.createJarFile(SOMELIB_JAR, SOMELIB_CLASSES);
+
+
+        // compile patch 1
+        compiled = CompilerUtils.compile(SOMELIB_PATCH1_SRC, SOMELIB_PATCH1_CLASSES,
+                        "--module-path", SOMELIB_JAR.toString(),
+                        "--add-modules", MODULE_NAME,
+                        "--patch-module", MODULE_NAME + "=" + SOMELIB_PATCH1_SRC);
+        assertTrue(compiled);
+
+        // compile patch 2
+        compiled = CompilerUtils.compile(SOMELIB_PATCH2_SRC, SOMELIB_PATCH2_CLASSES,
+                        "--module-path", SOMELIB_JAR.toString(),
+                        "--add-modules", MODULE_NAME,
+                        "--patch-module", MODULE_NAME + "=" + SOMELIB_PATCH2_SRC);
+        assertTrue(compiled);
+
+        // compile app
+        compiled = CompilerUtils.compile(APP_SRC, APP_CLASSES,
+                        "--module-path", SOMELIB_JAR.toString());
+        assertTrue(compiled);
+    }
+
+    @Test
+    public void modulePathExtend() throws Exception {
```

Comment:
I think it would be useful for future maintainers if there was a comment on each of the 4 tests. I'd probably rename them them too, e.g. testExtendAutomaticModuleOnModulePath, testExtendAutomaticModuleAsInitialModule, so that it's a bit clear what each one does.